sympy/solvers/solveset.py

```diff
@@ -2434,60 +2446,93 @@ def linear_coeffs(eq, *syms, **_kw):
+
+        >>> linear_coeffs(x**2, x)
+        Traceback (most recent call last):
+        ...
+        NonlinearError:
+        nonlinear in given generators
+
+        * unless the nonlinear term is given as a symbol or does not
+          contain, as a literal generator, one of the symbols given
+
+        >>> linear_coeffs(x + 2*x**2, x, x**2)
+        [1, 2, 0]
+        >>> linear_coeffs(x**4 + 2*x**2, x**2)
+        [2, x**4]
+
+    The last example is a feature of how the function works: ``x**4``
+    has, as a generator, ``x**2`` the same way ``x**2` as ``x`` as a
+    generator, but in the arg traversal, the `x**2`` is not seen in
+    ``x**4``.  More often, one will treat functions or derivatives
+    as independent of the symbols by using keyword `strict=False`:
+
+        >>> from sympy import cos
+        >>> eq, v = x*(cos(y) + 1) + 3, [x, y]
+        >>> linear_coeffs(eq, *v, strict=False)
+        [cos(y) + 1, 0, 3]
+
+    Regardless of the setting of `strict`, it should always be
+    true that the result will reconstruct an expression that is
+    equivalent to the input expression.
+
+        >>> from math import prod
+        >>> r = _  # result from above
+        >>> eq.equals(r[-1] + sum([prod(i) for i in zip(v, r)]))
+        True
     """
-    d = defaultdict(list)
+    from sympy.polys.matrices.linsolve import _lin_eq2dict
     eq = _sympify(eq)
+    if len(syms) == 1 and iterable(syms[0]) and not isinstance(syms[0], Basic):
+        raise ValueError('pass unpacked symbols, *syms')
     symset = set(syms)
     if len(symset) != len(syms):
         raise ValueError('duplicate symbols given')
-    has = set(iterfreeargs(eq)) & symset
-    if not has:
-        return [S.Zero]*len(syms) + [eq]
-    c, terms = eq.as_coeff_add(*has)
-    d[0].extend(Add.make_args(c))
-    for t in terms:
-        m, f = t.as_coeff_mul(*has)
-        if len(f) != 1:
-            break
-        f = f[0]
-        if f in symset:
-            d[f].append(m)
-        elif f.is_Add:
-            d1 = linear_coeffs(f, *has, **{'dict': True})
-            d[0].append(m*d1.pop(0))
-            for xf, vf in d1.items():
-                d[xf].append(m*vf)
-        else:
-            break
-    else:
-        for k, v in d.items():
-            d[k] = Add(*v)
-        if not _kw:
-            return [d.get(s, S.Zero) for s in syms]+ [d[0]]
-        return d  # default is still list but this won't matter
-    raise NonlinearError('nonlinear term encountered: %s' % t)
+    try:
+        c, d = _lin_eq2dict(eq, symset, strict)
+    except PolyNonlinearError as err:
+        raise NonlinearError(str(err)) from err
+    if dict:
+        if c:
+            d[1] = c
+        return d
+    rv = [S.Zero]*len(syms)
+    for i, k in enumerate(syms):
+        if k not in d:
+            continue
+        rv[i] = d[k]
+    return rv + [c]
 
 
-def linear_eq_to_matrix(equations, *symbols):
+def linear_eq_to_matrix(equations, *symbols, strict=True):
```

Comment:
I wonder here whether `strict` should default to True.

Reply:
Yes, the default is True.